### ISSUE TITLE
Proposal: break up documentation files into directories and files

### ISSUE THREAD

**Issue by vitorgalvao** [caskroom/homebrew-cask#16690]
I occasionally check [electron’s API documentation](https://github.com/atom/electron/tree/master/docs/api), and it’s a breeze. Not only can I go to that particular URL and immediately find a file that has *only* what I need (which also means <kbd>⌘</kbd><kbd>F</kbd> is much more useful), I can actually easily google for it.

Contrast that with our documentation, in which we have humongous files with an index at the top, and we need anchors to navigate.

My proposal: break it up. Specifically (and at least) our most important pieces of documentation: `CONTRIBUTING` and `CASK_LANGUAGE_REFERENCE`. Look at [electron’s documentation structure as a whole](https://github.com/atom/electron/tree/master/docs), and you’ll see how clear and easy to navigate it is.

Unless there are objections, I intend to start taking a crack at it.

**Comment by adityadalal924** [comment]
I actually am of the opposite opinion - I like having a single large file that I can search through. It's also really nice to have a *useful* CONTRIBUTING, which can be read straight through - I wouldn't be opposed to moving some parts of it out to `/doc, and leaving it as a barebones (but useful) guide.

I'll write up some more thoughts on CONTRIBUTING later - not currently at a computer.

For CASK_LANGUAGE_REFERENCE - it would probably be okay to split things up - I only ever use the *index* at the top, so if splitting it up makes your use case/navigation easier, then I say go for it.

tl;dr: YES to splitting up the language reference, NO to splitting up CONTRIBUTING (detailed thoughts on a restructure of that coming in a day or two)

**Comment by vitorgalvao** [comment]
> I wouldn't be opposed to moving some parts of it out to `/doc`, and leaving it as a barebones (but useful) guide.

I agree `CONTRIBUTING` benefits more than other documents from being a cohesive document you can read in succession, It’s also a bit harder to figure out where to split up, as opposed to `CASK_LANGUAGE_REFERENCE` where we can split up by stanza, but in my initial thougts for it, without going through it in depth, I was thinking mainly of (at least for starters) separating it into a few distinct sections:

+ Setting up to contribute
+ Contributing to the core
+ Updating a cask
+ Adding a cask

So the point would be to separate by intent. You likely only need `Setting up to contribute` to contribute once. After that, to make a change you only have one intention in mind, so once you’re inside the document for a particular context, your searching likely won’t be impaired.

**Comment by adityadalal924** [comment]
Getting Set Up To Contribute - keep inline
Updating a Cask - keep inline
Adding a Cask - link out to new doc
Testing Your New Cask - move out as part of adding a cask
Finding a Home For Your Cask - link out to new file
Submitting Your Changes - move out as part of adding a cask
Cleaning up - move out as part of adding a cask
Reporting Bugs - keep inline

Two things that don't have a great location - version helpers info and appcast info - maybe those need to be in LANGUAGE REFERENCE?

**Comment by tsparber** [comment]
I was just searching for the version helpers in LANGUAGE REFERENCE as I expected them there and found them in CONTRIBUTING only after searching for the pull request adding them. (And I even knew that table exists somewhere).

**Comment by adityadalal924** [comment]
@vitorgalvao Things might get messy if both of us work on this directly - let's figure out the final structure via discussion here, and then we can move forward with a PR. (I'll likely only be touching CONTRIBUTING though, so feel free to move forward with changes as desired to CASK_LANGUAGE_REFERENCE)

@tsparber I had the same thought, hence my point that they likely need a better spot (whether that be LANG REF or a new document, I have no preference)

**Comment by vitorgalvao** [comment]
`version` methods are indeed a tricky one. I don’t exactly know where they’d fit in `CASK_LANGUAGE_REFERENCE`. @tsparber You were looking for them there, in what section would you expect to have found them?

As for `appcast` info, even the details of how to add them I could see working inside `CASK_LANGUAGE_REFERENCE`.

@adityadalal924 Perhaps both documents should work more in unison. So for instance in `CONTRIBUTING` needing to update the `appcast` would be mentioned, with a link to the corresponding in-depth documentation in `CASK_LANGUAGE_REFERENCE`. Same with `style`, for example, as it is something you might want to look for on its own, and there’s really no need to have everything (like `stanza order`) inside `CONTRIBUTING`.

So I’m thinking of `CONTRIBUTING` as a bare-bones quick-start guide that links heavily to other documents. Kind of like a mix of a summary and an index.

**Comment by tsparber** [comment]
I searched for a "Version Stanza Details" section (like as the others). Basically containig the version methods section from contributing.

**Comment by adityadalal924** [comment]
@vitorgalvao CASK_LANGUAGE_REFERENCE could be left as an index, of sorts, linking out to individual markdown files (one per stanza). 

I also took another look at Electron's documentation and it really is nice - if over time, hacking.md started looking more like the "Development" section, that would be fantastic. 

CONTRIBUTING.md (the piece I'm currently working on) will likely end up acting like a Quick Start guide, with lots of links out to other files.

Are you also planning to add in more subdirectories to `/doc`?

**Comment by vitorgalvao** [comment]
> CASK_LANGUAGE_REFERENCE could be left as an index, of sorts, linking out to individual markdown files (one per stanza). 

Yes, I had something like that in mind.

> Are you also planning to add in more subdirectories to `/doc`?

Yes. Like electron’s documentation, I’d like to have a `README` in `/doc` that would serve as the index to the other stuff.

**Comment by jawshooah** [comment]
We could also use something like [readthedocs.org](https://readthedocs.org/) (see [SublimeLinter's documentation](http://www.sublimelinter.com/en/latest/) as an example).

**Comment by adidalal** [comment]
Also realized github has a wiki feature - it is a bit easier to edit web-based, at least (and is also git-backed, which is nice)
https://github.com/blog/699-making-github-more-open-git-backed-wikis

**Comment by adidalal** [comment]
Also checked out a few examples of readthedocs - it's rst instead of markdown, but the overall interface looks nicer and it's even easier for people to edit

@jawshooah good suggestion!

**Comment by jawshooah** [comment]
> it's rst instead of markdown

[It works with Markdown too!](it's rst instead of markdown) :tada:

**Comment by vitorgalvao** [comment]
I don’t think people read the wikis, though. All the ones I’ve found always felt a bit abandoned and less serious than documentation on the repo itself. Homebrew used to use the wiki, and they since switched to documents on the repo.

I don’t like readthedocs on account of its looks, to be honest (are there themes for it?). I honestly can never read documentation on it, as that sidebar is incredibly distracting. My eyes can only look at it, not at the actual content (not joking). I do like its system, though, and it should be easy to port our documentation to use it, if we decide to go that route.

I’m currently using [Gitbook](https://www.gitbook.com/)’s editor to break up the documentation.

**Comment by adidalal** [comment]
@vitorgalvao Something like http://fabric8.io/gitbook/?

**Comment by vitorgalvao** [comment]
@adidalal [They can indeed be connected](https://help.gitbook.com/github/index.html), as I understand it, just like readthedocs. However, it might not be able to hook into subdirectories, now that I look into it.

However, my last paragraph was very unclear. I am indeed using their editor, but that in no way ties with having to use their service. To clarify, I do think something like readthedocs is a good idea, I’d just like to have a bit more control/options regarding how the documentation is shown (damn sidebar).

Gitbook’s editor is nice because it allows to easily work with multiple markdown files at once. I’m actually taking suggestions for other editors that do the same, if anyone has them.

**Comment by adidalal** [comment]
I use Sublime Text 3 with Markdown Extended syntax highlighting and https://github.com/SublimeText-Markdown/MarkdownEditing
